src/frontend/src/planner/set_operation.rs

```diff
@@ -0,0 +1,67 @@
+use itertools::Itertools;
+// Copyright 2022 Singularity Data
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+use risingwave_common::error::{ErrorCode, Result};
+
+use crate::binder::{BoundSetExpr, BoundSetOperation};
+use crate::optimizer::plan_node::LogicalUnion;
+use crate::planner::Planner;
+use crate::PlanRef;
+
+impl Planner {
+    pub(super) fn plan_set_operation(
+        &mut self,
+        op: BoundSetOperation,
+        all: bool,
+        left: BoundSetExpr,
+        right: BoundSetExpr,
+    ) -> Result<PlanRef> {
+        match op {
+            BoundSetOperation::Union => {
+                let left = self.plan_set_expr(left, vec![])?;
+                let right = self.plan_set_expr(right, vec![])?;
+
+                if left.schema().fields.len() != right.schema().fields.len() {
+                    return Err(ErrorCode::InvalidInputSyntax(
+                        "each UNION query must have the same number of columns".to_string(),
+                    )
+                    .into());
+                }
+
+                for (a, b) in left
+                    .schema()
+                    .fields
+                    .iter()
+                    .zip_eq(right.schema().fields.iter())
+                {
+                    if a.data_type != b.data_type {
+                        return Err(ErrorCode::InvalidInputSyntax(format!(
+                            "UNION types {} of column {} is different from types {} of column {}",
+                            a.data_type.prost_type_name().as_str_name(),
+                            a.name,
+                            b.data_type.prost_type_name().as_str_name(),
+                            b.name,
+                        ))
+                        .into());
+                    }
+                }
```

Comment:
@xiangjinwu I have left schema check here. Currently, only accept the same schema of the both side of UNION.

Reply:
Yes I see this check in planner after reviewing the binder part.
* Any specific reason to delay the validation?
* Have you confirmed if the bind context issue is valid regarding `union order by a+1`?

Reply:
I have no convincing reason. I Just simply think the schema check is closer to the relational algebra than the parser or binder. Both are ok to me.

Reply:
Type checking (and casting) is part of the binder/validator. The boundary between binder and planner can be blur sometimes, but I usually view the former as answer to "what does this query mean" and the latter as a transformation from declarative to imperative (although still logical).

Reply:
Sure, I will move the checking from planner to binder later.